libnetwork/osl/interface_linux.go

```diff
@@ -276,22 +323,232 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
 	if err != nil {
 		return fmt.Errorf("failed to set link up: %v", err)
 	}
+	log.G(ctx).Debug("link has been set to up")
 
 	// Set the routes on the interface. This can only be done when the interface is up.
 	if err := setInterfaceRoutes(ctx, nlh, iface, i); err != nil {
 		return fmt.Errorf("error setting interface %q routes to %q: %v", iface.Attrs().Name, i.Routes(), err)
 	}
 
+	// Wait for the interface to be up and running (or a timeout).
+	up, err := waitForIfUpped(ctx, newNs, iface.Attrs().Index)
+	if err != nil {
+		return err
+	}
+
+	// If the interface is up, send unsolicited ARP/NA messages if necessary.
+	if up {
+		if err := n.advertiseAddrs(ctx, iface.Attrs().Index, i, nlh); err != nil {
+			return fmt.Errorf("failed to advertise addresses: %w", err)
+		}
+	}
+
 	n.mu.Lock()
 	n.iFaces = append(n.iFaces, i)
 	n.mu.Unlock()
 
 	return nil
 }
 
+func waitForIfUpped(ctx context.Context, ns netns.NsHandle, ifIndex int) (bool, error) {
+	update := make(chan netlink.LinkUpdate, 100)
+	upped := make(chan struct{})
+	opts := netlink.LinkSubscribeOptions{
+		ListExisting: true, // in case the link is already up
+		ErrorCallback: func(err error) {
+			select {
+			case <-upped:
+				// Ignore errors sent after the upped channel is closed, the netlink
+				// package sends an EAGAIN after it closes its netlink socket when it
+				// sees this channel is closed. (No message is ever sent on upped.)
+				return
+			default:
+			}
+			log.G(ctx).WithFields(log.Fields{
+				"ifi":   ifIndex,
+				"error": err,
+			}).Info("netlink error while waiting for interface up")
+		},
+	}
+	if ns.IsOpen() {
+		opts.Namespace = &ns
+	}
+	if err := nlwrap.LinkSubscribeWithOptions(update, upped, opts); err != nil {
+		return false, fmt.Errorf("failed to subscribe to link updates: %w", err)
+	}
+
+	timer := time.NewTimer(5 * time.Second)
+	defer timer.Stop()
+	defer close(upped)
+
+	for {
+		select {
+		case <-timer.C:
```

Comment:
```suggestion
		case <-time.After(time.Second * 5)
```

nit: With Go 1.23, there's no performance advantage of `time.NewTimer` over `time.After`, so we could just use the latter since it's a bit simpler.

Reply:
I've changed it to `time.After` - but created before the for-loop, so it doesn't create a new timer each time a message arrives on the `update` channel.